Translate Chapter 6 to PTBR (#12)
JuliaIntro/JuliaIntroBR.jl: book/src/chap06.asciidoc

> [[chap06]]

Reviewer: Fiz comentários no Código.

> == Funções produtivas

Reviewer: Camel Case: "Funções Produtivas"

Muitas das funções de Julia que usamos, como as funções matemáticas, produzem valores de retorno. Contudo, as funções que escrevemos são todas nulas: têm um efeito, como imprimir um valor ou mover uma tartaruga, mas retornam +nothing+. Neste capítulo você aprenderá a elaborar funções produtivas.
(((função produtiva)))

> === Valores de retorno

Reviewer: Camel Case

Author: Modified.

A chamada de uma função gera um valor de retorno, que geralmente atribuímos a uma variável ou utilizamos como parte de uma expressão.

[source,julia]
----
e = exp(1.0)
altura = raio * sin(radianos)
----

> As funções escritas até o momento são nulas, que significa que elas não têm valor de retorno; mais precisamente, seu valor de retorno é +nothing+. Neste capítulo, (finalmente) vamos escrever funções produtivas. O primeiro exemplo é a função +área+, que retorna a área de um círculo dado um raio:

Reviewer: ```suggestion
As funções escritas até o momento são nulas, o que significa que elas não têm valor de retorno; mais precisamente, seu valor de retorno é +nothing+. Neste capítulo, (finalmente) vamos escrever funções produtivas. O primeiro exemplo é a função +área+, que retorna a área de um círculo dado um raio:
```

Reviewer: Talvez o artigo "o" se encaixe bem

(((nothing)))(((área)))((("função", "definido pelo programador", "área", see="área")))

[source,@julia-setup]
----
function área(raio)
    var_área = π * raio^2

>     return var_área

Reviewer 2: ```suggestion
function área(raio)
    a = π * raio^2
    return a
```

end
----

> Já vimos a declaração +return+ anteriormente, mas em uma função produtiva a declaração +retorno+ inclui uma expressão. Esta declaração significa: "Retorne imediatamente a partir desta função e leve a expressão seguinte como um valor de retorno". Como esta expressão pode ser arbitrariamente complicada, então poderíamos ter escrito uma função mais sucinta:

Reviewer: Sugestão, em "Como **_esta_** expressão pode ser arbitrariamente complicada, então poderíamos ter escrito uma função mais sucinta:" trocar "esta" por "essa".
Esta --> Pra quando o objeto será referido,
Essa --> Quando o objeto já foi referido.

Author: Modified.

(((declaração de retorno)))

[source,@julia-setup]
----
function área(raio)
    π * raio^2
end
----

O valor retornado por uma função é o valor do último comando executado, que, por padrão, é a última expressão no corpo da definição da função.

> Além disso, _variáveis temporárias_ como +var_área+ e as intruções explícitas +return+ podem contribuir para o debugging.

Reviewer 2: ```suggestion
Além disso, _variáveis temporárias_ como +a+ e as instruções explícitas +return+ podem contribuir para o debugging.
```

> Além disso, _variáveis temporárias_ como +var_área+ e as intruções explícitas +return+ podem contribuir para a depuração.

Reviewer: Depuração = Debugação (?)

Author: Modified.

end
----

> E já que estas instruções de retorno estão em uma alternativa condicional, somente uma é executada.

Reviewer: E já que estas instruções de retorno estão em uma **_condicional alternativa_**, somente uma é executada.

(((alternativa condicional)))

> Assim que um comando +return+ é executado, a função termina sem executar qualquer comando posterior. O código que aparece após um comando +return+, ou qualquer outro lugar que o fluxo de execução não alcance, é chamado de _código morto_.

Reviewer: O código que aparece após um comando +return+, ou qualquer outro lugar que o fluxo de execução **_nunca possa alcançar_**.

end
----

> Essa função está incorreta porque se +x+ for 0, nenhuma das condições é verdadeira, e a função termina sem chegar a um comando de +retorno+. Quando o fluxo de execução chega ao final de uma função, o valor de retorno é +nothing+, que não é o valor absoluto de 0.

Reviewer 2: ```suggestion
Essa função está incorreta porque se +x+ for 0, nenhuma das condições é verdadeira, e a função termina sem chegar a um comando de +return+. Quando o fluxo de execução chega ao final de uma função, o valor de retorno é +nothing+, que não é o valor absoluto de 0.
```

Author: ```suggestion
Essa função está incorreta porque se +x+ for 0, nenhuma das condições é verdadeira, e a função termina sem chegar a um comando +return+. Quando o fluxo de execução chega ao final de uma função, o valor de retorno é +nothing+, que não é o valor absoluto de 0.
```

> Esta função está incorreta porque se +x+ for 0, nenhuma das condições é verdadeira, e a função termina sem chegar a um comando de +retorno+. Quando o fluxo de execução chegar ao final de uma função, o valor de retorno é +nothing+, que não é o valor absoluto de 0.

Reviewer: "**_Essa_** função está incorreta "

Reviewer: "Quando o fluxo de execução **_chega_** ao final de uma função, o valor de retorno é +nothing+, que não é o valor absoluto de 0."

> Esta função está incorreta porque se +x+ for 0, nenhuma das condições é verdadeira, e a função termina sem chegar a um comando de +retorno+. Quando o fluxo de execução chegar ao final de uma função, o valor de retorno é +nothing+, que não é o valor absoluto de 0.

Reviewer: **_Essa_** função está incorreta

[source,@julia-repl-test chap06]
----
julia> show(valorAbsoluto(0))
nothing
----

[TIP]
====

> Julia tem uma função interna chamada +abs+ que calcula valores absolutos.

Reviewer: "Julia **_possui_** uma função interna chamada"

(((abs)))((("função", "Base", "abs", see="abs")))
====

===== Exercício 6-1

Escreva uma função +comparar+ que recebe dois valores, +x+ e +y+, e retorna +1+ se +x > y+, +0+ se +x == y+ e +-1+ se +x < y+.

[[incremental_development]]

> === Desenvolvimento incremental

Reviewer: Camel Case

À medida que se escreve funções maiores, talvez aconteça de você passar mais tempo debugando.
(((debugging)))

> Ao lidar com programas cada vez mais complexos, tente um processo chamado _Desenvolvimento Incremental_, que adiciona e testa apenas uma pequena quantidade de código por vez, evitando assim longas sessões de debugging.

Reviewer 2: ```suggestion
Ao lidar com programas cada vez mais complexos, tente um processo chamado _desenvolvimento incremental_, que adiciona e testa apenas uma pequena quantidade de código por vez, evitando assim longas sessões de debugging.
```

> Ao lidar com programas cada vez mais complexos, tente um processo chamado _desenvolvimento incremental_, que adiciona e testa apenas uma pequena quantidade de código por vez, evitando assim longas sessões de debugging.

Reviewer: " evitando assim longas **_seções_** de debugging."

Seções = Corte
Sessão = Espaço de tempo. Ex.:Cinema

(((desenvolvimento incremental)))

> Para exemplificar este processo, suponha que você queira determinar a distância entre dois pontos, dada pelas coordenadas latexmath:[\(\left(x_1, y_1\right)\)] e latexmath:[\(\left(x_2, y_2\right)\)]. Pelo teorema de Pitágoras, a distância é:

Reviewer: "Para exemplificar **_esse_** processo"

[[incremental_development]]
=== Desenvolvimento Incremental

À medida que se escreve funções maiores, talvez aconteça de você passar mais tempo debugando.
(((debugging)))

> Ao lidar com programas cada vez mais complexos, tente um processo chamado _desenvolvimento incremental_, que adiciona e testa apenas uma pequena quantidade de código por vez, evitando assim longas sessões de debugging.

Reviewer: Em "_desenvolvimento incremental_" talvez deixar em Camel Case também

> Ao lidar com programas cada vez mais complexos, tente um processo chamado _desenvolvimento incremental_, que adiciona e testa apenas uma pequena quantidade de código por vez, evitando assim longas sessões de debugging.

Reviewer: Em "_desenvolvimento incremental_" talvez deixar em Camel Case também

Author: Fiz a alteração. Mas a versão em inglês não está em camel case, nem a versão da tradução em espanhol (Piense en Julia) e nem o Pense em Python.

\end{equation}
++++

> O primeiro passo é avaliar como deve ser uma função de distância em Julia. Em outras palavras, quais são as entradas (os parâmetros) e qual é a saída (o valor de retorno)?

Reviewer 2: ```suggestion
O primeiro passo é avaliar como deve ser uma função de distância em Julia. Em outras palavras, quais são as entradas (parâmetros) e qual é a saída (valor de retorno)?
```

Author: Ocultar os artigos é anglicismo.

Reviewer 2: :+1:

Neste exemplo, as entradas são dois pontos, que pode ser representado por quatro números. Já o valor de retorno é a distância representada por um valor de ponto flutuante.

Com isso, pode-se esboçar a função:

> (((distância))) (((("função", "definido pelo programador", "distância", see="distância"))))

Reviewer: ```suggestion
(((distância)))((("função", "definido pelo programador", "distância", see="distância")))
```

[source,@julia-setup chap06]
----
function distância(x₁, y₁, x₂, y₂)
    0.0
end
----

> Obviamente que esta versão não determina as distâncias pois ela sempre retorna zero. Mas é sintaticamente correta, e roda, o que significa que você pode testá-la antes de complicá-la. Os números dos subscritos estão disponíveis na codificação de caracteres Unicode (*+\_1 TAB+*, *+\_2 TAB+*, etc.).

Reviewer: **_Obviamente, essa_** versão não determina as distâncias pois ela sempre retorna zero.

> (((caracter Unicode)))

Reviewer 2: ```suggestion
(((caractere Unicode)))
```

Para testar a nova função, chame-a com argumentos exemplificados a seguir:

[source,@julia-repl-test chap06]
----
distância(1, 2, 4, 6)
----

> Escolhi estes valores para que a distância horizontal seja 3 e a vertical seja 4; logo, o resultado é 5 porque é a hipotenusa de um triângulo retângulo 3-4-5. Quando testar uma função, o aconselhável é já saber a resposta correta.

Reviewer: "**_Escolhi esses valores_** para que a distância"

> Escolhi estes valores para que a distância horizontal seja 3 e a vertical seja 4; logo, o resultado é 5 porque é a hipotenusa de um triângulo retângulo 3-4-5. Quando testar uma função, o aconselhável é já saber a resposta correta.

Reviewer 2: ```suggestion
Escolhemos estes valores para que a distância horizontal seja 3 e a vertical seja 4; logo, o resultado é 5 porque é a hipotenusa de um triângulo retângulo 3-4-5. Quando testar uma função, o aconselhável é já saber a resposta correta.
```

end
----

> Você executaria o programa mais uma vez nesse estágio e verificaria a saída (que deveria ser 25). Números sobrescritos também estão disponíveis (*+\^2 TAB+*). Por fim, usa-se +sqrt+ no calcular e retornar o resultado final:

Reviewer: Faltou traduzir "Again" para "Novamente"

> Você executaria o programa mais uma vez nesse estágio e verificaria a saída (que deveria ser 25). Números sobrescritos também estão disponíveis (*+\^2 TAB+*). Por fim, usa-se +sqrt+ no calcular e retornar o resultado final:

Reviewer 2: ```suggestion
Você executaria o programa mais uma vez nesse estágio e verificaria a saída (que deveria ser 25). Números sobrescritos também estão disponíveis (*+\^2 TAB+*). Por fim, usa-se +sqrt+ para calcular e retornar o resultado final:
```

(((sqrt)))

[source,@julia-setup]
----
function distância(x₁, y₁, x₂, y₂)
    dx = x₂ - x₁
    dy = y₂ - y₁
    d² = dx^2 + dy^2
    sqrt(d²)
end
----

Se a função rodar corretamente, pronto. Caso contrário, convém mostrar o valor de +sqrt(d²)+ antes do comando +return+.

> A versão final da função não exibe nada quando é executada, retornando apenas um valor. As instruções de impressão que escrevemos são úteis para o debugging, mas depois que a função estiver funcionando, devemos removê-las. Um código como esse é chamado _andaime_ porque é útil para criar o programa, embora não faça parte do produto final.

Reviewer: "que escrevemos são úteis para **_debugar_**,"

(((andaime)))

> Ao iniciar, você deve adicionar apenas uma ou duas linhas de código por vez. À medida que você ganha mais experiência, pode se escrever e debugar pedaços maiores. De qualquer forma, o desenvolvimento incremental pode economizar muito tempo de debugging.

Reviewer: "muito tempo de **_debug_**"

Os principais aspectos do processo são:

> . Comece com um programa funcional e faça pequenas alterações incrementais. A qualquer momento, se houver um erro, você deverá ter uma boa idéia de onde ele está.

Reviewer: "ter uma boa **_ideia_** de onde ele está"

. Use variáveis para armazenar valores intermediários de modo que você possa visualizá-los e conferi-los.

. Uma vez que o programa esteja funcionando, você pode querer retirar algumas das instruções andaimes ou consolidar múltiplos comandos em expressões compostas, desde que não dificulte a leitura do programa.

===== Exercício 6-2

Use o desenvolvimento incremental para escrever uma função chamada +hipotenusa+ que retorna o comprimento da hipotenusa de um triângulo retângulo, a partir dos comprimentos dos outros dois catetos como argumentos. Registre cada estágio do processo de desenvolvimento à medida que avança.

=== Composição

> Como já esperado, você pode chamar uma função de dentro da outra. Para exemplificar isto, escreveremos uma função que calcula a área do círculo a partir de dois pontos, o centro do círculo e um ponto no perímetro.

Reviewer: "Para exemplificar **_isso_**"

Isto - O que vai ser mencionado
Isso - Já mencionado anteriormente

(((composição)))

> Suponha que o ponto central é indicado pelas variáveis +xc+ e +yc+, e o ponto de perímetro indicado por +xp+ e +yp+. O primeiro passo é encontrar o raio do círculo, dado pela distância entre estes dois pontos. Note que acabamos de escrever a função distância:

Reviewer: "Suponha que o ponto central é indicado pelas variáveis +xc+ e +yc+, e **_que o_** ponto de perímetro **_é_** indicado por +xp+ e +yp+."

[source,julia]
----
raio = distância(xc, yc, xp, yp)
----

O próximo passo é calcular a área de um círculo a partir desse raio, e por isso também escrevemos esta função:

[source,julia]
----
resultado = área(raio)
----

Encapsulando esses passos em uma função, temos:
(((área_círculo)))((("função", "definido pelo programador", "área_círculo", see="área_círculo")))

[source,@julia-setup]
----
function área_círculo(xc, yc, xp, yp)
    raio = distância(xc, yc, xp, yp)
    resultado = área(raio)
    return resultado
end
----

As variáveis temporárias +raio+ e +resultado+ são úteis para o desenvolvimento e o debugging, mas depois que o programa estiver funcionando, podemos torná-lo mais conciso fazendo:
(((variável temporária)))

[source,@julia-setup]
----
function área_círculo(xc, yc, xp, yp)
    área(distância(xc, yc, xp, yp))
end
----

[[boolean_functions]]

> === Funções booleanas

Reviewer: Camel Case

> As funções podem retornar variáveis booleanas, o que muitas vezes é conveniente para ocultar testes complicados dentro de funções. Por exemplo:

Reviewer: "As funções podem retornar **_valores_** booleanas"

> (((função booleana))) (((é_divisível))) ((("função", "definido pelo programador", "é_divisível", consulte = "é_divisível")))

Reviewer: ```suggestion
(((função booleana)))(((é_divisível)))((("função", "definido pelo programador", "é_divisível", see="é_divisível")))
```

(((==)))

[source,@julia-setup]
----
function é_divisível(x, y)
    x % y == 0
end
----

Funções booleanas são constantemente utilizadas em estruturas condicionais:
(((estrutura condicional)))

[source,julia]
----
if é_divisível(x, y)
    println("x é divisível por y")
end
----

Talvez seja tentador escrever algo como:

[source,julia]
----
if é_divisível(x, y) == true
    println("x é divisível por y")
end
----

No entanto, a comparação adicional com +true+ é desnecessária.

===== Exercício 6-3

Escreva uma função +está_entre(x, y, z)+ que retorna +true+ se +x ≤ y ≤ z+ ou +false+ caso contrário.

=== Mais Recursividade

Mostramos apenas uma pequena fração de Julia, mas você pode estar interessado em saber que essa fração é uma linguagem de programação _completa_, significando que qualquer coisa que possa ser calculada pode ser expressa nessa linguagem. Qualquer programa já escrito pode ser reescrito usando apenas os recursos da linguagem que você aprendeu até o momento (na verdade, você precisaria de alguns comandos para controlar dispositivos como mouse, discos, etc., mas isso é tudo).

> (((recursão))) (((linguagem de programação completa))) ((("linguagem de programação", "completa", veja = "linguagem de programação completa"))))

Reviewer 2: ```suggestion
(((recursão)))(((linguagem de programação completa)))((("linguagem de programação", "completa", see="linguagem de programação completa")))
```

Essa afirmação é um exercício não trivial provado pela primeira vez por Alan Turing, um dos primeiros cientistas da computação (alguns argumentariam que ele era matemático, mas muitos dos primeiros cientistas da computação começaram como matemáticos). Por isso, esta prova é conhecida como a Tese de Turing. Para uma discussão mais completa (e precisa) da Tese de Turing, recomendo o livro _Introdução à teoria da computação_ de Michael Sipser.

> (((Turing, Alan))) (((tese de Turing))))

Reviewer: ```suggestion
(((Turing, Alan)))(((tese de Turing)))
```

Para ter uma noção do que você pode fazer com as ferramentas que sabe até agora, avaliaremos algumas funções matemáticas definidas recursivamente. Uma definição recursiva é semelhante a uma definição circular, no sentido de que a definição contém uma chamada de si própria. Uma definição totalmente circular não é muito vantajosa:

> (((definição recursiva))) (((definição circular))))

Reviewer: ```suggestion
(((definição recursiva)))(((definição circular)))
```

vorpal::
Um adjetivo usado para descrever algo que é vorpal.

Ver essa definição no dicionário pode ser irritante. Por outro lado, se consultar a definição da função fatorial, denotada com o símbolo latexmath:[\(!\)], poderá obter algo assim:
(((função fatorial)))

[latexmath]
++++
\begin{equation}
{n! =
\begin{cases}
  1& \textrm{se}\  n = 0 \\
  n (n-1)!& \textrm{se}\  n > 0
\end{cases}}
\end{equation}
++++

> Esta definição diz que o fatorial de 0 é 1, e o fatorial de qualquer outro valor latexmath:[n] é latexmath:[n] multiplicado pelo fatorial de latexmath:[n-1]].

Reviewer: "**_Essa_** definição "

> Essa definição diz que o fatorial de 0 é 1, e o fatorial de qualquer outro valor latexmath:[n] é latexmath:[n] multiplicado pelo fatorial de latexmath:[n-1]].

Reviewer 2: ```suggestion
Essa definição diz que o fatorial de 0 é 1, e o fatorial de qualquer outro valor latexmath:[\(n\)] é latexmath:[\(n\)] multiplicado pelo fatorial de latexmath:[\(n-1\)].
```

> Então latexmath:[\(3!\)] é 3 vezes latexmath:[\(2!\)], que é 2 vezes latexmath:[\(1!\)], que é 1 vezes latexmath:[\(0!\)]. Colocando tudo junto, látexmath:[\(3!\)] é igual a 3 vezes 2 vezes 1 vezes 1, que dá 6.

Reviewer 2: ```suggestion
Então latexmath:[\(3!\)] é 3 vezes latexmath:[\(2!\)], que é 2 vezes latexmath:[\(1!\)], que é 1 vezes latexmath:[\(0!\)]. Colocando tudo junto, latexmath:[\(3!\)] é igual a 3 vezes 2 vezes 1 vezes 1, que dá 6.
```

> Se puder escrever uma definição recursiva de algo, pode-se escrever um programa em Julia para testá-la. A primeiro etapa é decidir quais devem ser os parâmetros. E neste caso, é evidente que o fatorial recebe um número inteiro:

Reviewer: " E **_nesse_** caso,"

> Se puder escrever uma definição recursiva de algo, pode-se escrever um programa em Julia para testá-la. A primeiro etapa é decidir quais devem ser os parâmetros. E nesse caso, é evidente que o fatorial recebe um número inteiro:

Reviewer 2: ```suggestion
Se puder escrever uma definição recursiva de algo, pode-se escrever um programa em Julia para testá-la. A primeira etapa é decidir quais devem ser os parâmetros. E nesse caso, é evidente que o fatorial recebe um número inteiro:
```

> (((fatorial))) (((("função", "definido pelo programador", "fatorial", see="fatorial"))))

Reviewer: ```suggestion
(((fatorial)))((("função", "definido pelo programador", "fatorial", see="fatorial")))
```

Reviewer: Tem um parênteses a mais.

[source,@julia-setup]
----
function fatorial(n) end
----

Se o argumento for +0+, basta retornar +1+:

[source,@julia-setup]
----
function fatorial(n)
    if n == 0
        return 1
    end
end
----

Caso contrário, e esta é a parte interessante, temos que fazer uma chamada recursiva para encontrar o fatorial de +n-1+ para depois multiplicá-lo por +n+:

[source,@julia-setup]
----
function fatorial(n)
    if n == 0
        return 1
    else
        recursão = fatorial(n-1)
        resultado = n * recursão
        return resultado
    end
end
----

> O fluxo de execução deste programa é similar ao fluxo de +contagem regressiva+ em <<recursão>>. Chamando +fatorial+ do valor +3+:

Reviewer 2: ```suggestion
O fluxo de execução deste programa é similar ao fluxo de +contagem regressiva+ em <<recursion>>. Chamando +fatorial+ do valor +3+:
```

(((fluxo de execução)))

[small]
--
Como +3+ não é +0+, seguimos para o segundo ramo e calculamos o fatorial de +n-1+ ...

pass:[&#8193;]Como +2+ não é +0+, seguimos para o segundo ramo e calculamos o fatorial de +n-1+ ...

pass:[&#8193;&#8193;]Como +1+ não é +0+, seguimos para o segundo ramo e calculamos o fatorial de +n-1+ ...

pass:[&#8193;&#8193;&#8193;]Como +0+ é igual a +0+, seguimos para o primeiro ramo e temos o resultado +1+ sem efetuar pass:[<br/>&#8193;&#8193;&#8193;&#8193;]mais chamadas recursivas.

pass:[&#8193;&#8193;]O valor de retorno (= +1+) é multiplicado por +n+ (que é +1+), e o +resultado+ é devolvido.

pass:[&#8193;]O valor de retorno (= +1+), é multiplicado por +n+ (que é +2+), e o +resultado+ é devolvido.

> O valor de retorno +2+ é multiplicado por +n+ (que é +3+), e o resultado (= +6+), torna-se o valor de retorno da chamada da função que iniciou todo esse processo.

Reviewer: "O valor de retorno, **_(= +2+)_**," para manter o mesmo padrão acima

--

[[fig06-1]]

> .Stack diagram

Reviewer: Traduzir para "Diagrama de Pilha"

> .Stack diagram

Reviewer: Traduzir para "Diagrama de Pilha"

image::images/fig61.svg[]

<<fig06-1>> mostra como fica o diagrama de pilha para esta sequência de chamadas de função.
(((diagrama de pilha)))

Os valores de retorno são exibidos quando devolvidos de volta para cima da pilha. Em cada quadro, o valor de retorno é o valor de +resultado+, dado pelo produto de +n+ com +recursão+.

No último quadro, as variáveis locais +recursão+ e +resultado+ não existem porque o ramo que as cria não é executado.

[TIP]
====
Em Julia, a função +factorial+ calcula o fatorial de um número inteiro.

> (((factorial)))((("função", "Base", "factorial", see="factorial")))

Reviewer: Traduzir "factorial"

====

> === Salto de fé

Reviewer: Camel Case

Ler programas seguindo o fluxo de execução pode se tornar rapidamente exaustivo. Uma alternativa que eu chamo de "salto de fé" faz a leitura conforme o fluxo de execução e quando se chega a uma chamada de função, _assume_-se que a função funciona corretamente e devolve o resultado correto.
(((salto de fé)))

Na verdade, você já está praticando este salto de fé no uso de funções embutidas. Quando você chama +cos+ ou +exp+, você não investiga os corpos dessas funções. Você apenas assume que funcionam já que as pessoas que escreveram as funções embutidas eram bons programadores.

> A mesma prática ocorre quando você chama uma de suas próprias funções. Por exemplo, em <<funções booleanas>>, escrevemos a função +é_divisível+ que determina se um número é divisível por outro. Depois de nos convencermos de que essa função está correta ao examinar seu código e testar, podemos usá-la sem olhar para o corpo novamente.

Reviewer: Camel Case em "funções booleanas"

O mesmo se aplica aos programas recursivos. Ao chegar na chamada recursiva, em vez de acompanhar o fluxo de execução, deve-se assumir que a chamada recursiva funciona (retorna o resultado correto) e depois se perguntar: “Supondo que possa encontrar o fatorial do latexmath:[\(n-1\)], posso calcular o fatorial do latexmath:[\(n\)]?” Sim, multiplicando por latexmath:[\(n\)].

É claro que é um pouco estranho assumir que a função funciona corretamente quando ainda não se terminou de escrevê-la, mas é por isso que se chama salto de fé!

[[one_more_example]]
=== Mais Um Exemplo

Após fatorial, o exemplo mais familiar de uma função matemática definida recursivamente é a sequência de fibonacci, cuja definição é (consulte https://pt.wikipedia.org/wiki/Sequência_de_Fibonacci):

> (((função de fibonacci)))

Reviewer 2: ```suggestion
Após fatorial, o exemplo mais familiar de uma função matemática definida recursivamente é a sequência de Fibonacci, cuja definição é (consulte https://pt.wikipedia.org/wiki/Sequência_de_Fibonacci):
(((função de Fibonacci)))
```

[latexmath]
++++
\begin{equation}
{fib(n) =
\begin{cases}
    0& \textrm{se}\  n = 0 \\
    1& \textrm{se}\  n = 1 \\
    fib(n-1) + fib(n-2)& \textrm{se}\  n > 1
\end{cases}}
\end{equation}
++++

Traduzindo para Julia, tem-se:

> (((fib))) (((("função", "definido pelo programador", "fib", see="fib"))))

Reviewer: ```suggestion
(((fib)))((("função", "definido pelo programador", "fib", see="fib")))
```

====

=== Salto de Fé

> Ler programas seguindo o fluxo de execução pode se tornar rapidamente exaustivo. Uma alternativa que eu chamo de "salto de fé" faz a leitura conforme o fluxo de execução e quando se chega a uma chamada de função, _assume_-se que a função funciona corretamente e devolve o resultado correto.

Reviewer: Deixar em itálico em "_assume-se_"

(((salto de fé)))

Na verdade, você já está praticando este salto de fé no uso de funções embutidas. Quando você chama +cos+ ou +exp+, você não investiga os corpos dessas funções. Você apenas assume que funcionam já que as pessoas que escreveram as funções embutidas eram bons programadores.

> A mesma prática ocorre quando você chama uma de suas próprias funções. Por exemplo, em <<boolean_functions>>, escrevemos a função +é_divisível+ que determina se um número é divisível por outro. Depois de nos convencermos de que essa função está correta ao examinar seu código e testar, podemos usá-la sem olhar para o corpo novamente.

Reviewer: O que está em << >> não deveria ser traduzido?
Vou dar um double check na lista do Chris.

Author: Eu segui o que a tradução em espanhol fez e eles não traduzem as referências em << >>.

end
----

> Se tentar acompanhar o fluxo de execução aqui, mesmo para valores razoavelmente pequenos de +n+, sua cabeça vai enlouquecer. No entanto, de acordo com o salto de fé, se presumir que as duas chamadas recursivas funcionam sem erros, fica nítido que o resultado certo é obtido a partir da soma delas.

Reviewer 2: ```suggestion
Se você tentar acompanhar o fluxo de execução aqui, mesmo para valores razoavelmente pequenos de +n+, sua cabeça vai enlouquecer. No entanto, de acordo com o salto de fé, se presumir que as duas chamadas recursivas funcionam sem erros, fica nítido que o resultado certo é obtido a partir da soma delas.
```

> === Verificação de tipos

Reviewer: Camel Case

> O que ocorre se chamarmos +fatorial+ e atribuirmos +1,5+ como argumento?

Reviewer 2: ```suggestion
O que ocorre se chamarmos +fatorial+ e atribuirmos +1.5+ como argumento?
```

(((StackOverflowError)))

[source,jlcon]
----
julia> fatorial(1.5)
ERROR: StackOverflowError:
Stacktrace:
 [1] fatorial(::Float64) at ./REPL[3]:2
----

Parece uma recursão infinita. Como pode ser? A função tem um caso base—quando +n == 0+. Mas se +n+ não for um número inteiro, podemos _perder_ o caso base e ficar recursivo para sempre.
(((recursão infinita)))

> Na primeira chamada recursiva, o valor de +n+ é +0,5+. No próximo, é +-0,5+. A partir daí, vai diminuindo e ficando cada vez mais negativo, mas nunca será +0+.

Reviewer 2: ```suggestion
Na primeira chamada recursiva, o valor de +n+ é +0.5+. No próximo, é +-0.5+. A partir daí, vai diminuindo e ficando cada vez mais negativo, mas nunca será +0+.
```

Temos duas escolhas. Podemos tentar generalizar a função fatorial para trabalhar com números de ponto flutuante, ou podemos fazer +fatorial+ verificar o tipo de argumento. Na primeira opção, tem-se a função gama que está um pouco além do escopo deste livro. Logo, vamos adotar a segunda opção.
(((função gama)))

Podemos usar o operador embutido +isa+ para verificar o tipo do argumento. Ainda falando no assunto, também podemos certificar que o argumento seja positivo:
(((isa)))(((fatorial)))

[source,@julia-setup chap06]
----
function fatorial(n)
    if !(n isa Int64)
        error("Fatorial é definido somente para números inteiros.")
    elseif n < 0
        error("Fatorial não é definido para números inteiros negativos.")
    elseif n == 0
        return 1
    else
        return n * fatorial(n-1)
    end
end
----

> Enquanto o primeiro caso-base aborda os não-inteiros; o segundo aborda os inteiros negativos. Para estes dois casos, o programa exibe uma mensagem de erro e devolve +nothing+ para indicar que algo deu errado:

Reviewer: "Para **_esses_** dois casos"

[source,@julia-repl-test chap06]
----
julia> fatorial("fred")
ERROR: Fatorial é definido somente para números inteiros.
julia> fatorial(-2)
ERROR: Fatorial não é definido para números inteiros negativos.
----

Se passarmos pelas duas verificações, concluímos que +n+ é positivo ou zero, logo, conseguimos provar que a recursão termina.

> Este programa demonstra um padrão às vezes de _guardião_. Os dois primeiros condicionais atuam como guardiões, protegendo o código de valores que podem causar um erro. Além disso, os guardiões tornam possível provar a execução sem erro do código.

Reviewer: "**_Esse_** programa demonstra um padrão..."

        println(espaço, "retornando 1")
        return 1
    else
        recursão = fatorial(n-1)
        resultado = n * recursão
        println(espaço, "retornando ", resultado)
        return resultado
    end
end
----

+espaço+ é uma string de espaços que atua na indentação da saída:

[source,@julia-repl chap06]
----
fatorial(4)
----

Caso o fluxo de execução não esteja claro, esse tipo de saída de impressões pode ser útil. Leva algum tempo para usar andaimes eficientemente, mas um pouco de andaime pode economizar muito debugging.

=== Glossário

variável temporária::
Uma variável que armazena um valor intermediário em um cálculo difícil.
(((variável temporária)))

código morto::
O pedaço de um programa que nunca será executado, geralmente porque aparece após um comando de retorno.
(((código morto)))

desenvolvimento incremental::
Um plano de desenvolvimento de programa que tem o objetivo de evitar o debugging, adicionando e testando apenas uma pequena quantidade de código de cada vez.
(((desenvolvimento incremental)))(((plano de desenvolvimento de programa)))

andaime::
O código usado no decorrer do desenvolvimento do programa, porém não faz parte da versão final.

> ((((andaime)))

Reviewer: ```suggestion
(((andaime)))
```

Se passarmos pelas duas verificações, concluímos que +n+ é positivo ou zero, logo, conseguimos provar que a recursão termina.

Esse programa demonstra um padrão às vezes de _guardião_. Os dois primeiros condicionais atuam como guardiões, protegendo o código de valores que podem causar um erro. Além disso, os guardiões tornam possível provar a execução sem erro do código.
(((guardião)))

> Em <<catching_exceptions>>, veremos uma alternativa mais flexível para mostrar uma mensagem de erro: levantando uma exceção.

Reviewer: Traduzir o que está entre << >>, talvez para "Capturando Exceções". Tem que ver em qual capítulo isso deve aparecer.

Author: Eu segui o que a tradução em espanhol fez e eles não traduzem as referências em << >>.

----

[[ex06-2]]
===== Exercício 6-5

> Veja a função Ackermann, latexmath:[\(A(m, n)\)], definida como:

Reviewer: "Função **_de_** Ackermann"

(((função Ackermann)))

[latexmath]
++++
\begin{equation}
{A(m, n) =
\begin{cases}
              n+1& \textrm{se}\ m = 0 \\
        A(m-1, 1)& \textrm{se}\ m > 0\ \textrm{e}\ n = 0 \\
A(m-1, A(m, n-1))& \textrm{se}\ m > 0\ \textrm{e}\ n > 0.
\end{cases}}
\end{equation}
++++

> Consulte https://pt.wikipedia.org/wiki/Fun%C3%A7%C3%A3o_de_Ackermann. Escreva uma função chamada +ack+ que calcula a função Ackermann. Use sua função para avaliar +ack(3, 4)+, que é 125. O que ocorre quando aumentam os valores de +m+ e +n+?

Reviewer: "Função **_de_** Ackermann"

> ((((ack)))((("função", "definido pelo programador", "ack", see="ack")))

Reviewer: ```suggestion
(((ack)))((("função", "definido pelo programador", "ack", see="ack")))
```

[source,@julia-setup]
----
function primeira(palavra)
    primeira = firstindex(palavra)
    palavra[primeira]
end

function última(palavra)
    última = lastindex(palavra)
    palavra[última]
end

function meio(palavra)
    primeira = firstindex(palavra)
    última = lastindex(palavra)
    palavra[nextind(palavra, primeira) : prevind(palavra, última)]
end
----

Veremos como eles funcionam no <<chap08>>.

. Teste estas funções. O que acontece se você chamar meio para uma string de duas letras? E de uma letra? E no caso da string vazia, que é escrita +""+ e não tem nenhuma letra?

. Escreva uma função chamada +é_palíndromo+ que recebe um argumento string e retorna +true+ se for um palíndromo e +false+ caso contrário. Lembre-se de que você pode usar a função interna +length+ para verificar o comprimento de uma string.
(((é_palíndromo)))((("função", "definido pelo programador", "é_palíndromo", see="é_palíndromo")))(((comprimento)))

[[ex06-4]]
===== Exercício 6-7

> Um número, latexmath:[\(a\)], é dita uma potência do latexmath:[\(b\)] se for divisível por latexmath:[\(b\)] e latexmath:[\(\frac{a}{b}\)] é uma potência do latexmath:[\(b\)]. Escreva uma função chamada +é_potência+ que dados os parâmetros +a+ e +b+ devolve +true+ se +a+ for uma potência de +b+.

Reviewer 2: ```suggestion
Um número, latexmath:[\(a\)], é dito uma potência de latexmath:[\(b\)] se for divisível por latexmath:[\(b\)] e latexmath:[\(\frac{a}{b}\)] for potência de latexmath:[\(b\)]. Escreva uma função chamada +é_potência+ que dados os parâmetros +a+ e +b+ devolve +true+ se +a+ for uma potência de +b+.
```

> (((é_potência))) ((((função "," definido pelo programador ","é_potência", see="é_potência"))))

Reviewer: ```suggestion
(((é_potência)))((("função", "definido pelo programador","é_potência", see="é_potência")))
```

[TIP]
====
Você terá que considerar o caso base.
====

[[ex06-5]]
===== Exercício 6-8

O máximo divisor comum (MDC) de latexmath:[\(a\)] e latexmath:[\(b\)] é o maior número que divide os dois sem sobrar resto.

Uma maneira de encontrar o MDC de dois números é baseada na observação de que se latexmath:[\(r\)] é o resto da divisão de latexmath:[\(a\)] por latexmath:[\(b\)], então +mcd(a, b) = mcd(b, r)+. Para o caso base, considere que +mdc(a, 0) = a+.

> Escreva a função +mdc+ que recebe os parâmetros +a+ e +b+ e retorna o máximo divisor comum.

Reviewer: "Máximo Divisor Comum" em Camel Case

> (((mdc))) ((("função", "definido pelo programador", "mdc", consulte = "mdc"))))

Reviewer: ```suggestion
(((mdc)))((("função", "definido pelo programador", "mdc", see="mdc")))
```

> Crédito: Este exercício é baseado em um exemplo do livro _Structure and Interpretation of Computer Programs_ de Abelson e Sussman.

Reviewer 2: ```suggestion
Crédito: Este exercício é baseado em um exemplo do livro _Structure and Interpretation of Computer Programs_ de Abelson e Sussman.

```
